### DISCUSSION TITLE
Скорость прокрутки текста

### DISCUSSION THREAD

**Discussion by HuskySmile** [vvip-68/LedPanelWiFi#28] [Вопросы и ответы]
Добрый вечер. Имею матрицу 43х35 - все работает четко. Подскажите пожалуйста возможно ли увеличить скорость прокрутки текста, ползунок выкручен на максимум (253 больше не ставится), но текст как-то медленно идет? Хотя на второй матрице 30х16 с теме-же настройками текст летает на скорости 253.

**Comment by vvip-68** [comment]
Думаю что нет. Скорость зависит от размеров матрицы. Когда экран сформирован - он отправляет пакет импульсов (текущий экран) на матрицу. И пока отправка не закончится - микроконтроллер больше ничего не делает. Чем больше диодов - тем длиннее пакет, тем больше времени нужно на его отправку. При экране 16x16 пакет маленький и экраны отправляются за меньшее время - больше циклов в секунду. При 43x35 отправка экрана требует больше времени - меньше циклов в секунду.

После того как отправка пакета завершена, на максимальной скорости текста - формируется следующий экран со сдвинутым на одну колонку текстом. Единственный вариант в данном случае - делать каждый сдвиг не на одну колонку, а, скажем, сразу на две.

Сдвиг выполняется в runningText.ino, функция называется shiftTextPosition()

**Reply by vvip-68** [reply]
Еще возможный вариант решения - можно попробовать разбить матрицу на сегменты (до 4 в данной версии скетча), каждый сегмент настроить на свою ногу микроконтроллера. Тогда вывод на сегменты матрицы будет идти параллельно. Теоретически это должно увеличить частоту обновления экрана и, соответственно, увеличить скорость бегущей строки.

**Reply by huskydogsmile** [reply]
Понятно, спасибо за советы буду пробывать

**Reply by HuskySmile** [reply]
Второй способ сработал отлично, изначально текст с отсчетом до нового года проходил за 18 сек. Разбил матрицу на 3 сегмента и теперь прокрутка составляет 8 сек. Но появились артефакты в виде вспышек, но это дело второе, думаю скорее всего из-за увеличения длины сигнальных проводов.
Подскажите пожалуйста чтобы испытать первый способ нужно просто вписать цифру 2 между скобок shiftTextPosition(2) ?
извиняюсь что я не программист и не соображаю как сделать сдвиг на 2 колонки....

**Reply by vvip-68** [reply]
Не так. Внутри функции есть `offset--;`. Вместо него написать `offset -= 2;`
А если длинные провода скрутить косичкой?

**Reply by vvip-68** [reply]
Судя по размерам матрицы - у вас ESP32.
Нет ли после нескольких минут работы (около 10) фризов на некоторых эффектах?
В соседней ветке жалуются, что эффекты превращаются в слайд-шоу с очень низким fps.
А в логе начинают появляться сообщения из ядра микроконтроллера о каких-то ошибках с сетью...

**Reply by huskydogsmile** [reply]
Да esp 32. Работает все без перебоев и с фпс проблем никаких нет. 
Была проблема - перезагружался мк на некоторых эффектах. Долго думал почему, а решилось банально заменой флешки, хорошо что 2 разные купил. Так бы до сих пор репу чесал)

**Reply by HuskySmile** [reply]
Добрый день. В общем с этой бегущей строкой целая история) Удалось мне избавиться от вспышек.
Но проблема возникла в другом. Матрица 43х15, разделил на 3 сегмента 14х35, 14х35, 15х35
К примеру вот этот текст  {B#000000}До Нового года осталось {R01.01.***+ 0:00#N}  идеально по скорости прокрутки и без фризов (8.5 секунд - эталон)))), 
на этом {B#000000}До Нового года осталось {R01.01.***+}{S01.12.****#31.12.**** 23:59:59"} уже появляются рывки и мелкие фризы, простыми словами как будто что-то притормаживает и время уже 10.5 секунд
вот эта строка {B#000000} В городе {WS} {WT} °C {D}  аналогично с рывками и фризами
а на этой {B#000000}Сегодня {D:D} {D:MMMM} {D:DDDD} {C#000001}{D} еще хуже становится уже заметнее рывки с попытками ускориться). Пустой текст без макросов прокручивается тоже с мелкими фризами. 
Если поделить на 2 сегмента матрицу, фризов нет все плавно, но скорость не та))), не 8.5 сек а уже 12.5

Если есть какие мысли по этому поводу буду рад выслушать

В общем намучался я с этим текстом и все-же придется наверное остановиться на 2х сегментах.

**Reply by vvip-68** [reply]
Ну, можно попробовать с разными версиями ядра esp32 - 1.0.6, 2.0.6, 2.0.14... Возможно они себя будут вести как-то по-разному при выводе на параллельные сегменты ...

**Reply by vvip-68** [reply]
А вариант сдвига строки не на одну колонку, а сразу на две - не рассматривали? Это увеличит скорость прокрутки строки в два раза.

**Reply by HuskySmile** [reply]
Этот вариант тоже пробовал, но текст начинает читаться не четко, а с мерцанием чтоль, не знаю как объяснить. В общем глазу не приятно...

**Reply by saippuakauppias** [reply]
Про пакеты вообще интересная история на самом деле. Бегущая строка двигается не особо прям уж быстро, там меняется немного диодов. Но я залил несколько сконвертированных гифок, где очень много диодов меняют цвета – и там это происходит гораздо быстрее, чем бежит строка ) Хотя пакет данных явно больше уходит с МК

**Reply by shkolin73** [reply]
Это странно, но действительно бегущая строка очень по разному смотрится на разных эффектах. На некоторых она уж очень прерывисто смотрится. Мне понравилось на эффекте №43 - Звёзды.
